### PR TITLE
query: Introduce v1alpha2, allowing for multiple filters

### PR DESCRIPTION
**Pull request by adamwg**
### Description of your changes

To make the query API more flexible, while retaining simplicity for simple queries, allow multiple filters to be specified in a query. Objects returned in the query response must match all criteria in at least one of the specified filters; i.e., within a given filter the criteria are a conjunction (AND), while the multiple filters are a disjunction (OR). This is a breaking change, hence the version bump to v1alpha2.

I have:

- [x] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [x] Run `make reviewable` to ensure this PR is ready for review.
~- [ ] Added `backport release-x.y` labels to auto-backport this PR, as appropriate.~

### How has this code been tested

Manual testing in local dev.

### CODE REVIEW

File: apis/query/v1alpha2/query_types.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+	// objects specifies what to filter. Objects in the query response will
+	// match all criteria in at least one of the specified filters.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	Objects []QueryFilter `json:"objects"`
+}
+
+// ObjectIDs returns a flat list of object IDs specified in any of the top-level
+// filter's object filters.
+func (tlf *QueryTopLevelFilter) ObjectIDs() []string {
```
Comment (adamwg) [inline apis/query/v1alpha2/query_types.go:54]: @sttts Per your comment on #98 I've updated this to return `[]string`. Previously I had an issue with the consumer of this code and returned `[]any` as a shortcut, but I've sorted that out.

**Comment by adamwg** [comment]
Rebased to integrate the change from #108 into v1alpha2.

File: apis/query/v1alpha2/query_types.go [hunk 2]
```diff
@@ -75,49 +88,58 @@ type Freshness struct {
...
-// A QueryFilter specifies what to filter.
+// QueryFilter specifies what to filter. Objects in the query response will
+// match all criteria specified in the filter.
 type QueryFilter struct {
-	// namespace is the namespace WITHIN a control plane to query. If empty,
-	// all namespaces are queried in the given scope.
+	// id is the object ID to query.
+	ID string `json:"id,omitempty"`
+	// createdAt queries for objects with a range of creation times.
+	CreatedAt QueryCreatedAt `json:"createdAt,omitempty"`
```
Comment (sttts) [inline apis/query/v1alpha2/query_types.go:97]: Is CreatedAt really of interest? I could imagine more an interval.

File: apis/query/v1alpha2/query_types.go [hunk 3]
```diff
@@ -75,49 +88,58 @@ type Freshness struct {
...
-	// given scope.
-	Group string `json:"group,omitempty"`
-	// kind is the API kind to query. If empty, all kinds are queried in the
-	// given scope. The kind is case-insensitive. The kind also matches plural
-	// resources.
-	Kind string `json:"kind,omitempty"`
-	// categories is a list of categories to query. If empty, all categories are
-	// queried in the given scope.
+	// kind is the GroupKinds of objects to query.
+	Kind QueryGroupKind `json:"kind,omitempty"`
```
Comment (sttts) [inline apis/query/v1alpha2/query_types.go:103]: Call it GroupKind? A kind is a kind, not a GroupKind.

File: apis/query/v1alpha2/query_types.go (hunk 2 above)
Comment (sttts) [reply apis/query/v1alpha2/query_types.go:97]: oh, it is an interval. Maybe `creationTimestamp` is better?
```yaml
creationTimestamp:
  before: "..."
```

File: apis/query/v1alpha2/query_types.go [hunk 4]
```diff
@@ -75,49 +88,58 @@ type Freshness struct {
...
+type QueryCreatedAt struct {
+	After  metav1.Time `json:"after,omitempty"`
+	Before metav1.Time `json:"before,omitempty"`
 }
 
-// QueryFilterControlPlane specifies how to filter objects by control plane.
+// QueryGroupKind specifies how to query for GroupKinds.
+type QueryGroupKind struct {
+	// group is the API group to query. If empty all groups will be queried.
+	Group string `json:"group,omitempty"`
```
Comment (sttts) [inline apis/query/v1alpha2/query_types.go:129]: we use apiGroup elsewhere.

File: apis/query/v1alpha2/query_types.go [hunk 5]
```diff
@@ -75,49 +88,58 @@ type Freshness struct {
...
 type QueryFilterControlPlane struct {
 	// name is the name of the control plane to query. If empty, all control planes
 	// are queried in the given scope.
 	Name string `json:"name,omitempty"`
-	// namespace is the namespace of the control plane to query. If empty, all
-	// namespaces are queried in the given scope.
-	Namespace string `json:"namespace,omitempty"`
+	// group is the group of the control plane to query. If empty, all groups
+	// are queried in the given scope.
+	Group string `json:"group,omitempty"`
```
Comment (sttts) [inline apis/query/v1alpha2/query_types.go:142]: below for condition, maybe also reason? It is meant to be machine-readable.

File: apis/query/v1alpha2/query_types.go [hunk 6]
```diff
@@ -42,10 +42,23 @@ type QueryTopLevelFilter struct {
 	// control planes are queried in the given scope.
 	ControlPlane QueryFilterControlPlane `json:"controlPlane,omitempty"`
 
-	//	# ids: ["id1","id2"] # to get objects explicitly by id.
-	IDs []string `json:"ids,omitempty"`
+	// objects specifies what to filter. Objects in the query response will
+	// match all criteria in at least one of the specified filters.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	Objects []QueryFilter `json:"objects"`
```
Comment (sttts) [inline apis/query/v1alpha2/query_types.go:49]: No need for minlength IMO.

File: apis/query/v1alpha2/query_types.go (hunk 2 above)
Comment (adamwg) [reply apis/query/v1alpha2/query_types.go]: Agreed - `createdAt` doesn't sound like a range. Updated to `creationTimestamp`.

File: apis/query/v1alpha2/query_types.go (hunk 5 above)
Comment (adamwg) [reply apis/query/v1alpha2/query_types.go:141]: Makes sense - added. I removed the `required` validation from `Status` in the condition as well - not sure it ever made sense, but definitely doesn't make sense when `Reason` can also be specified. The semantics will be:
* if neither `Status` nor `Reason` is specified, any object with a condition of the given `Type` is matched.
* if `Status` and/or `Reason` are specified, the condition with the given `Type` must also have the right status and/or reason to match.

File: apis/query/v1alpha2/query_types.go (hunk 6 above)
Comment (sttts) [reply apis/query/v1alpha2/query_types.go:48]: neither of `+required` :)

**Review by sttts — APPROVED** [review]

**Review by phisco — APPROVED** [review]
LGTM 🙏 thanks @adamwg

File: apis/query/v1alpha2/query_types.go [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+	// filter specifies how to filter the returned objects.
+	Filter QueryTopLevelFilter `json:"filter,omitempty"`
+}
+
+// QueryNestedResources specifies how to return nested resources.
+type QueryNestedResources struct {
+	QueryResources `json:",inline"`
+
+	// filters specifies how to filter the returned objects.
+	Filters []QueryFilter `json:"filter,omitempty"`
```
Comment (luxas) [inline apis/query/v1alpha2/query_types.go:243]: filters

Comment (adamwg) [reply apis/query/v1alpha2/query_types.go:243]: Good catch! Pretty sure no one's using v1alpha2 yet, so should be safe to fix it. I'll put up a PR.

**Pull request merged**